Conceptual doc for platform compat analyzer (#20688)
dotnet/docs: docs/standard/analyzers/platform-compat-analyzer.md

---
title: Platform compatibility analyzer
description: A Roslyn analyzer can help detect platform compatibility issues across cross platform .NET 5.0.
author: buyaa-n
ms.date: 09/17/2020
---

# Platform compatibility analyzer

> By now everybody must heard of the mojo of `One .NET` - a single unified platform for building any type of application. The `.NET 5.0` SDK includes ASP.NET Core, Entity Framework Core, WinForms, WPF, Xamarin and ML.NET and we will continue to gain support for more platforms over time. While we strive to provide an experience where you don't have to reason about the different kinds of .NET, we still don't want to fully abstract away the underlying OS, so you'll continue to be able to call OS specific APIs, be that via P/Invokes, WinRT, or the Xamarin bindings for iOS and Android.

Reviewer: ```suggestion
You've probably heard the motto of "One .NET": a single, unified platform for building any type of application. The .NET 5.0 SDK includes ASP.NET Core, Entity Framework Core, WinForms, WPF, Xamarin, and ML.NET, and will add support for more platforms over time. .NET 5.0 strives to provide an experience where you don't have to reason about the different flavors of .NET, but doesn't attempt to fully abstract away the underlying operating system (OS). You'll continue to be able to call OS-specific APIs, for example, P/Invokes, WinRT, or the Xamarin bindings for iOS and Android.
```

> But using platform-dependent APIs on a component makes the code no longer work across all platforms. We needed a way to detect this at design-time so that developers get diagnostics when they inadvertently use platform-specific APIs. To achieve this goal we introduced [platform compatibility analyzer](/visualstudio/code-quality/ca1416) and complementary APIs help developers identify and use platform-specific APIs where appropriate, the new APIs include:

Reviewer: ```suggestion
But using platform-dependent APIs on a component means the code no longer works across all platforms. We needed a way to detect this at design time so developers get diagnostics when they inadvertently use platform-specific APIs. To achieve this goal, .NET 5.0 introduces the [platform compatibility analyzer](/visualstudio/code-quality/ca1416) and complementary APIs to help developer identify and use platform-specific APIs where appropriate. 
The new APIs include:
```

> - Added <xref:System.Runtime.Versioning.SupportedOSPlatformAttribute> to annotate APIs as being platform-specific and <xref:System.Runtime.Versioning.UnsupportedOSPlatformAttribute> to express that and API is unsupported on a particular OS, optionally including the version numbers. These attributes have been applied to platform-specific APIs in the core .NET libraries (still in progress).

Reviewer: ```suggestion
- <xref:System.Runtime.Versioning.SupportedOSPlatformAttribute> to annotate APIs as being platform-specific and <xref:System.Runtime.Versioning.UnsupportedOSPlatformAttribute> to annotate APIs as being unsupported on a particular OS. These attributes can optionally include the version number, and have already been applied to some platform-specific APIs in the core .NET libraries.
```

> - Added `Is<Platform>()` and `Is<Platform>VersionAtLeast(int major, int minor = 0, int build = 0, int revision = 0)` style static methods in the <xref:System.OperatingSystem?displayProperty=nameWithType> class for safely calling platform dependent APIs, for example, <xref:System.OperatingSystem.IsWindows?displayProperty=nameWithType> could be used for guarding a call of windows specific APIs and  <xref:System.OperatingSystem.IsWindowsVersionAtLeast%2A?displayProperty=nameWithType>() could be used for guarding a versioned windows specific API call. Learn [more](#the-platform-specific-api-calls-in-the-above-examples-guarded-with-guard-methods) about how they used as guards of platform-specific API references.

Reviewer: ```suggestion
- `Is<Platform>()` and `Is<Platform>VersionAtLeast(int major, int minor = 0, int build = 0, int revision = 0)` static methods in the <xref:System.OperatingSystem?displayProperty=nameWithType> class for safely calling platform-specific APIs. For example, <xref:System.OperatingSystem.IsWindows?displayProperty=nameWithType> can be used to guard a call to a Windows-specific API, and <xref:System.OperatingSystem.IsWindowsVersionAtLeast%2A?displayProperty=nameWithType>() can be used to guard a versioned Windows-specific API call. See these [examples](#the-platform-specific-api-calls-in-the-above-examples-guarded-with-guard-methods) of how these methods can be used as guards of platform-specific API references.
```

> > [!Tip]

Reviewer: ```suggestion
> [!TIP]
```

> > Platform compatibility analyzer upgrades/replaces [Discovering cross-platform issues](../../standard/analyzers/api-analyzer.md#discover-cross-platform-issues) feature of the [.NET API analyzer](../../standard/analyzers/api-analyzer.md).

Reviewer: ```suggestion
> The platform compatibility analyzer upgrades and replaces the [Discovering cross-platform issues](../../standard/analyzers/api-analyzer.md#discover-cross-platform-issues) feature of the [.NET API analyzer](../../standard/analyzers/api-analyzer.md).
```

## Prerequisites

> Platform compatibility analyzer is one of the Roslyn code quality analyzers. Starting from .NET 5.0, these analyzers are [included with the .NET SDK](../../fundamentals/productivity/code-analysis.md). For older .Net versions you can install code quality analyzers from [NuGet package](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers). The platform compatibility analyzer is enabled by default only for projects targeting `net5.0` or later. However, you can [enable](/visualstudio/code-quality/ca1416.md#configurability) it for projects targeting other frameworks.

Reviewer: ```suggestion
The platform compatibility analyzer is one of the Roslyn code quality analyzers. Starting in .NET 5.0, these analyzers are [included with the .NET SDK](../../fundamentals/productivity/code-analysis.md). For older .NET versions, you can install code quality analyzers from [NuGet package](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers). The platform compatibility analyzer is enabled by default only for projects that target `net5.0` or a later version. However, you can [enable](/visualstudio/code-quality/ca1416.md#configurability) it for projects that target other frameworks.
```

Does the NuGet package include the platform compatibility analyzer?

Author: I don't see NuGet package for 5.0, older versions wouldn't have it, so maybe better remove this `NuGet package` section

> ### How the analyzer detect platform dependency using the `SupportedOSPlatform` and `UnsupportedOSPlatform` attributes?

Reviewer: ```suggestion
### How the analyzer detects platform dependency using the `SupportedOSPlatform` and `UnsupportedOSPlatform` attributes
```

- An **unattributed API** is considered to work on **all OS platforms**.
- An API marked with `[SupportedOSPlatform("platform")]` is considered only portable to the specified OS `platform`.
  - The attribute can be applied multiple times to indicate **multiple platform support** (`[SupportedOSPlatform("windows"), SupportedOSPlatform("Android6.0")]`).

>   - The analyzer would produce **warning** if platform-specific APIs referenced without a proper **platform context** such as:

Reviewer: ```suggestion
  - The analyzer will produce a **warning** if platform-specific APIs are referenced without a proper **platform context**:
```

>     - **Warns** if the project is not targeting the supported platform (for example for windows-specific API call the project is targeting `<TargetFramework>net5.0-ios14.0</TargetFramework>`).

Reviewer: ```suggestion
    - **Warns** if the project doesn't target the supported platform (for example, a Windows-specific API call and the project targets `<TargetFramework>net5.0-ios14.0</TargetFramework>`).
```

    - **Warns** if the project is multi-targeted (`<TargetFramework>net5.0</TargetFramework>`).

>     - **Not warns** if the platform-specific API is referenced within a project targeting any of the **specified platforms** (for example for windows-specific API call the project is targeting `<TargetFramework>net5.0-windows</TargetFramework>`).

Reviewer: ```suggestion
    - **Doesn't warn** if the platform-specific API is referenced within a project that targets any of the **specified platforms** (for example, for a Windows-specific API call and the project targets `<TargetFramework>net5.0-windows</TargetFramework>`).
```

>     - **Not warns** if the platform specific API call is guarded by corresponding platform check methods (for example as within `if(OperatingSystem.IsWindows())` condition).

Reviewer: ```suggestion
    - **Doesn't warn** if the platform-specific API call is guarded by corresponding platform-check methods (for example, `if(OperatingSystem.IsWindows())`).
```

>     - **Not warns** if the `platform` specific API is referenced from the same `platform` specific context (**call site also attributed** with `[SupportedOSPlatform("platform")`).

Reviewer: ```suggestion
    - **Doesn't warn** if the platform-specific API is referenced from the same platform-specific context (**call site also attributed** with `[SupportedOSPlatform("platform")`).
```

> - An API marked with `[UnsupportedOSPlatform("platform")]` is considered unsupported only to the specified OS `platform` but supported for all other platforms.

Reviewer: ```suggestion
- An API marked with `[UnsupportedOSPlatform("platform")]` is considered unsupported only on the specified OS `platform` but supported for all other platforms.
```

>   - The attribute can be applied multiple times with different platforms. (`[UnsupportedOSPlatform("iOS"), UnsupportedOSPlatform("Android6.0")]`).

Reviewer: ```suggestion
  - The attribute can be applied multiple times with different platforms, for example, `[UnsupportedOSPlatform("iOS"), UnsupportedOSPlatform("Android6.0")]`.
```

>   - The analyzer could produce **warning** only if the `platform` is effective for the call site, such as:

Reviewer: ```suggestion
  - The analyzer produces a **warning** only if the `platform` is effective for the call site:
```

>     - **Warns** if the project is targeting the `platform` attributed as unsupported (for example if the API attributed with `[UnsupportedOSPlatform("browser")]` and call site targeting `<TargetFramework>net5.0-browser</TargetFramework>`).

Reviewer: ```suggestion
    - **Warns** if the project targets the platform that's attributed as unsupported (for example, if the API is attributed with `[UnsupportedOSPlatform("browser")]` and the call site targets `<TargetFramework>net5.0-browser</TargetFramework>`).
```

>     - **Warns** if the project is multi-targeted and the `platform` is included in the default [MSBuild `<SupportedPlatform>`](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SupportedPlatforms.props) items group or the `platform` manually included within the `MSBuild` \<SupportedPlatform> items group:

Reviewer: ```suggestion
    - **Warns** if the project is multi-targeted and the `platform` is included in the default [MSBuild `<SupportedPlatform>`](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SupportedPlatforms.props) items group, or the `platform` is manually included within the `MSBuild` \<SupportedPlatform> items group:
```

    ```XML
    <ItemGroup>
        <SupportedPlatform Include="platform" />
    </ItemGroup>
    ```

>     - **Not warns** if you're building an app that is not targeting the unsupported `platform` or it is multi-targeted and the `platform` is not included in the default [MSBuild `<SupportedPlatform>`](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SupportedPlatforms.props) items group.

Reviewer: ```suggestion
    - **Doesn't warn** if you're building an app that doesn't target the unsupported platform or is multi-targeted and the platform is not included in the default [MSBuild `<SupportedPlatform>`](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SupportedPlatforms.props) items group.
```

- Both attributes can be instantiated with or without version numbers as part of the platform name.
  - Version numbers are in the format of `major.minor[.build[.revision]]`; `major.minor` is required and the `build` and `revision` parts are optional. For example, "Windows7.0" indicates Windows version 7.0, but "Windows" is interpreted as Windows 0.0.

> - Platform attributes can be applied to `Types, Members (methods, fields, properties and events) and Assembly`.

Reviewer: ```suggestion
- Platform attributes can be applied to types, members (methods, fields, properties, and events) and assemblies.
```

>   - Attributes applied at the top level `target` affects into all `Members and/or Types` inside it.

Reviewer: ```suggestion
  - Attributes applied at the top level `target` affect all of its members and types.
```

> - Same `platform` attribute can be applied across the symbol tree for different symbols only with different versions. (Example: assume an `Assembly` having a `[SupportedOSPlatform("windows")]` attribute and there is a new `Type` added with `windows 10.0.17763.0` release, that `Type` should be annotated with `[SupportedOSPlatform("window10.0.17763.0")]`).

Reviewer: ```suggestion
- The same `platform` attribute can be applied across the symbol tree for different symbols only with different versions. For example, an assembly has a `[SupportedOSPlatform("windows")]` attribute and there is a new type added with `windows 10.0.17763.0` release; that type should be annotated with `[SupportedOSPlatform("window10.0.17763.0")]`.
```

>   - In case `[SupportedOSPlatform("platformVersion")]` applied more than twice for the referenced member with same `platform` only Min and Max versions will be accounted for the analyzer.

Reviewer: ```suggestion
  - If `[SupportedOSPlatform("platformVersion")]` is applied more than twice for the referenced member with the same `platform`, only the minimum and maximum versions are considered by the analyzer.
```

>   - In case `[UnsupportedOSPlatform("platformVersion")]` applied more than twice for the referenced member with same `platform` only Min and next Min versions will be accounted for the analyzer.

Reviewer: ```suggestion
  - If `[UnsupportedOSPlatform("platformVersion")]` is applied more than twice for the referenced member with the same `platform`, only the too earliest versions are considered by the analyzer.
```

> See [examples of how the attributes work and what diagnostics they cause](#examples-of-how-the-attributes-work-and-what-diagnostics-they-produce)

Reviewer: ```suggestion
For more information, see [examples of how the attributes work and what diagnostics they cause](#examples-of-how-the-attributes-work-and-what-diagnostics-they-produce).

```

> ### Advanced scenarios for attributes combination

Reviewer: ```suggestion
### Advanced scenarios for combining attributes
```

> - If a combination of `[SupportedOSPlatform] and [UnsupportedOSPlatform]` attributes are present, we group all attributes by OS platform identifier:

Reviewer: ```suggestion
- If a combination of `[SupportedOSPlatform]` and `[UnsupportedOSPlatform]` attributes are present, all attributes are grouped by OS platform identifier:
```

>   - **Allow list**. If the lowest version for each OS platform is a `[SupportedOSPlatform]` attribute, the API is considered to only be supported by the listed platforms and unsupported by all other platforms. The optional `[UnsupportedOSPlatform]` attributes for each platforms could only have higher version of the Min supported version, which denotes that the API is removed starting from the specified version.

Reviewer: ```suggestion
  - **Allow list**. If the lowest version for each OS platform is a `[SupportedOSPlatform]` attribute, the API is considered to only be supported by the listed platforms and unsupported by all other platforms. The optional `[UnsupportedOSPlatform]` attributes for each platform can only have higher version of the minimum supported version, which denotes that the API is removed starting from the specified version.
```

  ```csharp
  // The API only supported on windows 8.0 and later, not supported for all other
  // The API is removed/unsupported from version 10.0.19041.0
  [SupportedOSPlatform("windows8.0")]
  [UnsupportedOSPlatform("windows10.0.19041.0")]
  public void ApiSupportedFromWindows80SupportFromCertainVersion();
  ```

>   - **Deny list**. If the lowest version for each OS platform is an `[UnsupportedOSPlatform]` attribute, then the API is considered to only be unsupported by the listed platforms and supported by all other platforms. The list could have `[SupportedOSPlatform]` attribute with same platform but only with higher version which denotes that the API is added support from that version.

Reviewer: ```suggestion
  - **Deny list**. If the lowest version for each OS platform is an `[UnsupportedOSPlatform]` attribute, then the API is considered to only be unsupported by the listed platforms and supported by all other platforms. The list could have `[SupportedOSPlatform]` attribute with the same platform but a higher version, which denotes that the API is supported starting from that version.
```

  
  ```csharp
  // The API was unsupported on windows until version 10.0.19041.0
  // The API is considered supported everywhere else without constraints.
  [UnsupportedOSPlatform("windows")]
  [SupportedOSPlatform("windows10.0.19041.0")]
  public void ApiSupportedFromWindows8UnsupportFromWindows10();
  ```

>   - **Inconsistent list**. If the lowest version for some platforms is `[SupportedOSPlatform]` while it is `[UnsupportedOSPlatform]` for other platforms, is considered inconsistent, which is not supported for the analyzer. We plan to introduce another analyzer producing a warning in case of inconsistency in the future.

Reviewer: ```suggestion
  - **Inconsistent list**. If the lowest version for some platforms is `[SupportedOSPlatform]` while it is `[UnsupportedOSPlatform]` for other platforms, is considered inconsistent, which is not supported for the analyzer. 
```

Docs shouldn't talk about future plans.

>   - If the lowest versions of `[SupportedOSPlatform]` and `[UnsupportedOSPlatform]` attributes are equal the analyzer would count it as **Deny list**.

Reviewer: ```suggestion
  - If the lowest versions of `[SupportedOSPlatform]` and `[UnsupportedOSPlatform]` attributes are equal, the analyzer considers the platform as part of the **Deny list**.
```

  
  > [!NOTE]

>   > An API that was supported initially and unsupported (removed) later versions are not expected to get supported back

Reviewer: ```suggestion
  > An API that was supported initially but unsupported (removed) in a later version is not expected to get resupported in an even later version.
```

      // warns:'StartedWindowsSupportFromVersion8' is supported on 'windows' 8.0 and later
      StartedWindowsSupportFromVersion8();

      // warns:'StartedWindowsSupportFrom8UnsupportedFrom10' is unsupported on 'windows'  
      // warns:'StartedWindowsSupportFrom8UnsupportedFrom10' is supported on 'windows' 8.0 and later
      // even there were 3 diagnostics found analyzer warn only for the first 2
      StartedWindowsSupportFrom8UnsupportedFrom10();
  }
  ```

### How to handle the warnings reported?

The recommend way to deal with these diagnostics is by making sure you only call these APIs when running on the appropriate platforms. You have five options on how you can address these warnings, choose which ever appropriate for your situation.

1. **Guard the call**. You can either achieve this by excluding the code at build time using `#if` and multi-targeting or by conditionally calling the code at runtime. Check whether you’re running on desired `Platform` by using any of platform check methods, like `OperatingSystem.Is<Platform>()` or `OperatingSystem.Is<Platform>VersionAtLeast(int major, int minor = 0, int build = 0, int revision = 0)`.

> 2. **Mark the call site as platform-specific**. You can also choose to mark your own APIs as being platform-specific, thus effectively just forwarding the requirements to your callers. You can mark the containing method or type or the entire assembly with same attributes as of the referenced platform-dependent call. [Examples](#call-site-marked-as-platform-specific-for-platform-specific-api-calls-in-above-examples)

Reviewer: ```suggestion
- **Mark the call site as platform-specific**. You can also choose to mark your own APIs as being platform-specific, thus effectively just forwarding the requirements to your callers. Mark the containing method or type or the entire assembly with the same attributes as the referenced platform-dependent call. [Examples](#call-site-marked-as-platform-specific-for-platform-specific-api-calls-in-above-examples).
```

> 3. **Assert the call site with platform check**. If you don't want the overhead of an additional if statement at run time, use <xref:System.Diagnostics.Debug.Assert(System.Boolean)?displayProperty=nameWithType>. [Example](#assert-the-call-site-with-platform-check)

Reviewer: ```suggestion
- **Assert the call site with platform check**. If you don't want the overhead of an additional `if` statement at run time, use <xref:System.Diagnostics.Debug.Assert(System.Boolean)?displayProperty=nameWithType>. [Example](#assert-the-call-site-with-platform-check).
```

> 4. **Delete the code**. Generally not what you want because it means you lose fidelity when your code is used by Windows users, but for cases where a cross-platform alternative exists, you’re likely better off using that over platform-specific APIs.

Reviewer: ```suggestion
- **Delete the code**. Generally not what you want because it means you lose fidelity when your code is used by Windows users. For cases where a cross-platform alternative exists, you’re likely better off using that over platform-specific APIs.
```

> 5. **Suppress the warning**. You can of course cheat and simply suppress the warning, either via editor.config or #pragma warning disable. However, you should prefer options (1), (2) and (3) when using platform-specific APIs.

Reviewer: ```suggestion
- **Suppress the warning**. You can also simply suppress the warning, either via editor.config or `#pragma warning disable`. However, this option should be a last resort when using platform-specific APIs.
```

> ### The platform specific API calls in the above examples guarded with guard methods

Reviewer: ```suggestion
### Guard platform-specific APIs with guard methods
```

> Guard methods platform name should match with the calling platform dependent API platform name. If platform string of the calling API include version:

Reviewer: ```suggestion
The guard method's platform name should match with the calling platform-dependent API platform name. If the platform string of the calling API includes the version:
```

> - For `[SupportedOSPlatform("platformVersion")]` attribute, guard method platform `version` should be >= calling platform `Version`

Reviewer: ```suggestion
- For the `[SupportedOSPlatform("platformVersion")]` attribute, the guard method platform `version` should be greater than or equal to the calling platform's `Version`.
```

> - For `[UnsupportedOSPlatform("platformVersion")]` attribute, guard method platform `version` should be <= calling platform `Version`

Reviewer: ```suggestion
- For the `[UnsupportedOSPlatform("platformVersion")]` attribute, the guard method's platform `version` should be less than or equal to the calling platform's `Version`.
```

      {
          SupportedOnWindowsAndLinux(); // not warn
      }

      // Can use &&, || logical operators to guard combined attributes
      if (OperatingSystem.IsWindowsVersionAtLeast(8) && !OperatingSystem.IsWindowsVersionAtLeast(10.0.19041)))
      {
          ApiSupportedFromWindows8UnsupportFromWindows10();
      }

      if (OperatingSystem.IsWindowsVersionAtLeast(10, 0, 1903))
      {
          AssemblySupportedOnWindowsApiSupportedFromWindows10(); // Only need to check latest supported version
      }
  }

  public void CallingUnsupportedApis() // Deny list calls
  {
      if (!OperatingSystem.IsAndroid())
      {
          DoesNotWorkOnAndroid(); // no diagnostic
      }

      if (!OperatingSystem.IsWindows() || OperatingSystem.IsWindowsVersionAtLeast(8))
      {
          StartedWindowsSupportFromVersion8(); // no diagnostic
      }

      if (!OperatingSystem.IsWindows() || // supported all other platform
         (OperatingSystem.IsWindowsVersionAtLeast(8) && !OperatingSystem.IsWindowsVersionAtLeast(10, 0, 1903)))
      {
          StartedWindowsSupportFrom8UnsupportedFrom10(); // no diagnostic
      }
  }
  ```

> ### Call site marked as platform specific for platform specific API calls in above examples

Reviewer: ```suggestion
### Mark call site as platform-specific
```

> Platform names should match with the calling platform dependent API. If platform string include version:

Reviewer: ```suggestion
Platform names should match the calling platform-dependent API. If the platform string includes a version:
```

> - For `[SupportedOSPlatform("platformVersion")]` attribute, the call site platform `version` should be >= calling platform `Version`

Reviewer: Please make same changes suggested in the previous section for these two bullet points.

---
title: Platform compatibility analyzer
description: A Roslyn analyzer that can help detect platform compatibility issues in cross-platform .NET 5.0 apps.
author: buyaa-n
ms.date: 09/17/2020
---

# Platform compatibility analyzer

You've probably heard the motto of "One .NET": a single, unified platform for building any type of application. The .NET 5.0 SDK includes ASP.NET Core, Entity Framework Core, WinForms, WPF, Xamarin, and ML.NET, and will add support for more platforms over time. .NET 5.0 strives to provide an experience where you don't have to reason about the different flavors of .NET, but doesn't attempt to fully abstract away the underlying operating system (OS). You'll continue to be able to call OS-specific APIs, for example, P/Invokes, WinRT, or the Xamarin bindings for iOS and Android.

> But using platform-dependent APIs on a component means the code no longer works across all platforms. We needed a way to detect this at design time so developers get diagnostics when they inadvertently use platform-specific APIs. To achieve this goal, .NET 5.0 introduces the [platform compatibility analyzer](/visualstudio/code-quality/ca1416) and complementary APIs to help developer identify and use platform-specific APIs where appropriate. 

Reviewer: ```suggestion
But using platform-dependent APIs on a component means the code no longer works across all platforms. We needed a way to detect this at design time so developers get diagnostics when they inadvertently use platform-specific APIs. To achieve this goal, .NET 5.0 introduces the [platform compatibility analyzer](/visualstudio/code-quality/ca1416) and complementary APIs to help developers identify and use platform-specific APIs where appropriate.
```

> [!TIP]
> The platform compatibility analyzer upgrades and replaces the [Discovering cross-platform issues](../../standard/analyzers/api-analyzer.md#discover-cross-platform-issues) feature of the [.NET API analyzer](../../standard/analyzers/api-analyzer.md).

## Prerequisites

The platform compatibility analyzer is one of the Roslyn code quality analyzers. Starting in .NET 5.0, these analyzers are [included with the .NET SDK](../../fundamentals/productivity/code-analysis.md). For older .NET versions, you can install code quality analyzers from [NuGet package](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers). The platform compatibility analyzer is enabled by default only for projects that target `net5.0` or a later version. However, you can [enable](/visualstudio/code-quality/ca1416.md#configurability) it for projects that target other frameworks.

> ### How the analyzer detects platform dependency using the `SupportedOSPlatform` and `UnsupportedOSPlatform` attributes

Reviewer: ```suggestion
## How the analyzer determines platform dependency
```

    - **Doesn't warn** if the platform-specific API is referenced from the same platform-specific context (**call site also attributed** with `[SupportedOSPlatform("platform")`).
- An API marked with `[UnsupportedOSPlatform("platform")]` is considered unsupported only on the specified OS `platform` but supported for all other platforms.
  - The attribute can be applied multiple times with different platforms, for example, `[UnsupportedOSPlatform("iOS"), UnsupportedOSPlatform("Android6.0")]`.
  - The analyzer produces a **warning** only if the `platform` is effective for the call site:
    - **Warns** if the project targets the platform that's attributed as unsupported (for example, if the API is attributed with `[UnsupportedOSPlatform("browser")]` and the call site targets `<TargetFramework>net5.0-browser</TargetFramework>`).
    - **Warns** if the project is multi-targeted and the `platform` is included in the default [MSBuild `<SupportedPlatform>`](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SupportedPlatforms.props) items group, or the `platform` is manually included within the `MSBuild` \<SupportedPlatform> items group:

>     ```XML

Reviewer: This code snippet should be indented two more spaces to align with the list.

For more information, see [examples of how the attributes work and what diagnostics they cause](#examples-of-how-the-attributes-work-and-what-diagnostics-they-produce).

### Advanced scenarios for combining attributes

- If a combination of `[SupportedOSPlatform]` and `[UnsupportedOSPlatform]` attributes are present, all attributes are grouped by OS platform identifier:
  - **Allow list**. If the lowest version for each OS platform is a `[SupportedOSPlatform]` attribute, the API is considered to only be supported by the listed platforms and unsupported by all other platforms. The optional `[UnsupportedOSPlatform]` attributes for each platform can only have higher version of the minimum supported version, which denotes that the API is removed starting from the specified version.

    ```csharp
    // The API only supported on Windows 8.0 and later, not supported for all other.
    // The API is removed/unsupported from version 10.0.19041.0.
    [SupportedOSPlatform("windows8.0")]
    [UnsupportedOSPlatform("windows10.0.19041.0")]
    public void ApiSupportedFromWindows80SupportFromCertainVersion();
    ```

>   - **Deny list**. If the lowest version for each OS platform is an `[UnsupportedOSPlatform]` attribute, then the API is considered to only be unsupported by the listed platforms and supported by all other platforms. The list could have `[SupportedOSPlatform]` attribute with the same platform but a higher version, which denotes that the API is supported starting from that version.

Author: Agree, for me "allow" and "deny" list is not that clear term, but i didn't wanted to change the original design definition of @terrajobst

>   ```csharp

Reviewer: This code snippet and the one on lines 64-70 should be indented two more spaces to align with the list.

  // The API was unsupported on windows until version 10.0.19041.0
  // The API is considered supported everywhere else without constraints.
  [UnsupportedOSPlatform("windows")]
  [SupportedOSPlatform("windows10.0.19041.0")]
  public void ApiSupportedFromWindows8UnsupportFromWindows10();
  ```

>   - **Inconsistent list**. If the lowest version for some platforms is `[SupportedOSPlatform]` while it is `[UnsupportedOSPlatform]` for other platforms, is considered inconsistent, which is not supported for the analyzer. 

Reviewer: ```suggestion
  - **Inconsistent list**. If the lowest version for some platforms is `[SupportedOSPlatform]` while it is `[UnsupportedOSPlatform]` for other platforms, is considered inconsistent, which is not supported for the analyzer.
```

  [UnsupportedOSPlatform("windows")]
  [SupportedOSPlatform("windows8.0")]
  public void StartedWindowsSupportFromVersion8() { }

  // an API was unsupported on windows until version8.0.
  // Then the API is removed (unsupported) from version 10.0.19041.0.
  // The API is considered supported everywhere else without constraints.
  [UnsupportedOSPlatform("windows")]
  [SupportedOSPlatform("windows8.0")]
  [UnsupportedOSPlatform("windows10.0.19041.0")]
  public void StartedWindowsSupportFrom8UnsupportedFrom10() { }

  public void Caller2()
  {
      DoesNotWorkOnAndroid(); // warns 'DoesNotWorkOnAndroid' is unsupported on 'android'

      // warns:'StartedWindowsSupportFromVersion8' is unsupported on 'windows'  
      // warns:'StartedWindowsSupportFromVersion8' is supported on 'windows' 8.0 and later
      StartedWindowsSupportFromVersion8();

      // warns:'StartedWindowsSupportFrom8UnsupportedFrom10' is unsupported on 'windows'  
      // warns:'StartedWindowsSupportFrom8UnsupportedFrom10' is supported on 'windows' 8.0 and later
      // even there were 3 diagnostics found analyzer warn only for the first 2
      StartedWindowsSupportFrom8UnsupportedFrom10();
  }
  ```

> ### Handle reported warnings

Reviewer: ```suggestion
## Handle reported warnings
```

The recommended way to deal with these diagnostics is to make sure you only call platform-specific APIs when running on an appropriate platform. Following are the options you can use to address the warnings; choose whichever is most appropriate for your situation:

> - **Guard the call**. You can either achieve this by excluding the code at build time using `#if` and multi-targeting or by conditionally calling the code at run time. Check whether you’re running on a desired `Platform` by using one of platform-check methods, for example, `OperatingSystem.Is<Platform>()` or `OperatingSystem.Is<Platform>VersionAtLeast(int major, int minor = 0, int build = 0, int revision = 0)`.

Reviewer 2: We should separate the `#if` and `OperatingSystem` approaches. They're quite distinct. And I recommend we move the `#if` approach to the end, as it requires defining custom MSBuild constants per target.

- **Mark the call site as platform-specific**. You can also choose to mark your own APIs as being platform-specific, thus effectively just forwarding the requirements to your callers. Mark the containing method or type or the entire assembly with the same attributes as the referenced platform-dependent call. [Examples](#mark-call-site-as-platform-specific).

- **Assert the call site with platform check**. If you don't want the overhead of an additional `if` statement at run time, use <xref:System.Diagnostics.Debug.Assert(System.Boolean)?displayProperty=nameWithType>. [Example](#assert-the-call-site-with-platform-check).

- **Delete the code**. Generally not what you want because it means you lose fidelity when your code is used by Windows users. For cases where a cross-platform alternative exists, you’re likely better off using that over platform-specific APIs.

> - **Suppress the warning**. You can also simply suppress the warning, either via editor.config or `#pragma warning disable`. However, this option should be a last resort when using platform-specific APIs.

Reviewer 2: I recommend using `#pragma warning disable ca1416` to promote disabling the single diagnostic.

### Guard platform-specific APIs with guard methods

The guard method's platform name should match with the calling platform-dependent API platform name. If the platform string of the calling API includes the version:

- For the `[SupportedOSPlatform("platformVersion")]` attribute, the guard method platform `version` should be greater than or equal to the calling platform's `Version`.
- For the `[UnsupportedOSPlatform("platformVersion")]` attribute, the guard method's platform `version` should be less than or equal to the calling platform's `Version`.

  ```csharp
  public void CallingSupportedOnlyApis() // Allow list calls
  {
      if (OperatingSystem.IsWindows())
      {
          WindowsOnlyApi(); // not warn
      }

      if (OperatingSystem.IsLinux())
      {

>           SupportedOnWindowsAndLinuxOnly(); // nor warn, within one of the supported context

Reviewer 2: Typo: "nor" should be "not"

      }

>       if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) // traditional method

Reviewer 2: "traditional" is a weird term. We should downplay this API but have a section for if you need to guard code that targets netstandard or netcoreapp as well as net5.0.  In those cases, this API is useful and will be respected. But it's not as optimized.

      {

>           SupportedOnWindowsAndLinux(); // not warn

Reviewer 2: I recommend "will not warn" or "no warning" for these comments

      }

      // Can use &&, || logical operators to guard combined attributes
      if (OperatingSystem.IsWindowsVersionAtLeast(8) && !OperatingSystem.IsWindowsVersionAtLeast(10.0.19041)))
      {
          ApiSupportedFromWindows8UnsupportFromWindows10();
      }

      if (OperatingSystem.IsWindowsVersionAtLeast(10, 0, 1903))
      {
          AssemblySupportedOnWindowsApiSupportedFromWindows10(); // Only need to check latest supported version
      }
  }

>   public void CallingUnsupportedApis() // Deny list calls

Reviewer 2: I think we can omit the comment here (and avoid the deny/allow terms)

  {
      if (!OperatingSystem.IsAndroid())
      {

>           DoesNotWorkOnAndroid(); // no diagnostic

Reviewer 2: Maybe "no warning" (whichever term we pick, they should be consistent)

      }

      if (!OperatingSystem.IsWindows() || OperatingSystem.IsWindowsVersionAtLeast(8))
      {
          StartedWindowsSupportFromVersion8(); // no diagnostic
      }

>       if (!OperatingSystem.IsWindows() || // supported all other platform

Reviewer 2: "platforms"

  [UnsupportedOSPlatform("windows")]
  [SupportedOSPlatform("windows8.0")]
  public void StartedWindowsSupportFromVersion8() { }

  // an API was unsupported on Windows until version8.0.
  // Then the API is removed (unsupported) from version 10.0.19041.0.
  // The API is considered supported everywhere else without constraints.
  [UnsupportedOSPlatform("windows")]
  [SupportedOSPlatform("windows8.0")]
  [UnsupportedOSPlatform("windows10.0.19041.0")]
  public void StartedWindowsSupportFrom8UnsupportedFrom10() { }

  [UnsupportedOSPlatform("windows")] // Caller no support Windows for any version.
  public void Caller4()
  {
      DoesNotWorkOnAndroid(); // warns 'DoesNotWorkOnAndroid' is unsupported on 'Android'.

      // not warns as the call site not support Windows at all, but supports all other.
      StartedWindowsSupportFromVersion8();

      // same, not warns as the call site not support Windows at all, but supports all other.
      StartedWindowsSupportFrom8UnsupportedFrom10();
  }

  [UnsupportedOSPlatform("windows")]
  [UnsupportedOSPlatform("android")] // Caller not support Windows and Android for any version.
  public void Caller4()
  {

>       DoesNotWorkOnAndroid(); // not warn as call site not supports Android.

Reviewer 2: "not warn" could be improved throughout here

    ```

  - **Inconsistent list**. If the lowest version for some platforms is `[SupportedOSPlatform]` while it is `[UnsupportedOSPlatform]` for other platforms, is considered inconsistent, which is not supported for the analyzer.
  - If the lowest versions of `[SupportedOSPlatform]` and `[UnsupportedOSPlatform]` attributes are equal, the analyzer considers the platform as part of the **Supported only list**.
- Platform attributes can be applied to types, members (methods, fields, properties, and events) and assemblies with different platform name and / or version.
  - Attributes applied at the top level `target` affect all of its members and types.

>   - Child level attributes only apply if they adhere to the rule `child annotations can narrow the platforms support, but they cannot widen it`.

Reviewer 2: ```suggestion
  - Child level attributes only apply if they adhere to the rule "child annotations can narrow the platforms support, but they cannot widen it".
```
